### ISSUE TITLE
Reducing max_steps slows or prevents learning

### ISSUE THREAD

**Issue by HPlumley** [Unity-Technologies/ml-agents#1568]
I have a simple discrete navigation environment, similar to the GridWorld example.

I used the training hyperparameters from Grid World and it was able to successfully train in just under 40k steps out of the 500k max steps. Improvement was minimal between 40k and 80k steps and flatlined after that.

I therefore tried reducing max_steps to 50k, however that causes the agent to completely fail to learn. It quickly ends up getting stuck in a policy where it walks into the corner and bounces off it.

I tried fiddling with other hyperparameters to recover the rapid learning of the long run but it becomes very unstable, sometimes reaching a successful policy in under 15k steps and sometimes not at all.

What other hyperparameters should be changed, and how, to have the same results as a run with a larger max_steps?

**Comment by MarcoMeter** [comment]
By changing max_steps, you alter the behavior of the learning rate, because the learning rate linearly decays based on the remaining steps.

So with 500k steps, the learning rate shrinks down much slower in comparison to 50k steps.

This is basically what's going on.

Maybe increase the batch size, epochs or learning rate.

**Comment by HPlumley** [comment]
Thanks for the tips Marco!

I have tried increasing learning rate, but as it decays linearly that fails to recover the success of the longer run.

If there are no other differences between a short and long run, is it possible to change the behaviour of the learning rate decay? Maybe rapid learning could be possible with a constant initial learning rate that then declines after a set number of steps.

Yes I have noticed how unstable it can be. But with some sets of parameters it can be quite stable too. For instance the long run always approaches the maximum reward in between 30k and 50k steps (tested with 10 runs, terminating after it reaches that point).

**Comment by JPhilipp** [comment]
> By changing max_steps, you alter the behavior of the learning rate, because
> the learning rate linearly decays based on the remaining steps.

Jumping in here, does this mean when I use a really big max_steps -- just so that it runs longer without me having to restart it -- it would actually decrease relative learning speed?

Differently put, is there a difference between running
  max_steps = 50k  * started 10 times
vs
  max_steps = 500k * started 1 times
?

**Comment by MarcoMeter** [comment]
@HPlumley 
You'd have to modify this line in order to alter the behavior of the learning rate:
https://github.com/Unity-Technologies/ml-agents/blob/master/ml-agents/mlagents/trainers/ppo/models.py#L167

@JPhilipp 
Do you mean starting 10 different training runs, there each one trains for 50k steps?

**Comment by JPhilipp** [comment]
> Do you mean starting 10 different training runs, there each one trains for 50k steps?

@MarcoMeter Yes exactly -- and that in comparison to starting 1 training run at 500k steps. So my question is, would the end result for these two situations be pretty much the same, or is it something very different (maybe because of the learning rate factor)? I had figured it would be the same, but what I read in this thread made me wonder.

**Comment by MarcoMeter** [comment]
If you continue the training of one model, I would expect the same behavior as you were training straight for 500k steps.

**Comment by JPhilipp** [comment]
Thanks!

**Comment by HPlumley** [comment]
I'd have to say no to that being equivalent @JPhilipp and @MarcoMeter. Unless you adjust the learning rate behaviour to decay by only 10% each time you ran it for 50k steps.

Having the rate go from 3.0e-4 to 0 ten times is a very different situation to a single straight line from 3.0e-4 to 0.

@MarcoMeter 
Thanks for those links, I'll take a look at changing how the learning rate decays to find the best solution.

**Comment by JPhilipp** [comment]
@HPlumley Thanks! So just to understand, if I have trained a brain at say max_steps: 100000, and it's on the right way but I feel needs another training round, would I then simply set max_steps: 200000 and run "mlagents-learn config/trainer_config.yaml --run-id=MyBrain --train --load"? Or is that not the sensible thing to do? (For what it's worth, that's the approach I'm currently using, and it sort of works, but I'm not sure it's the most effective as per the learning rate behavior.)
